parallelize live point proposals

joshspeagle opened the pull request:
This is a slightly modified version of @aarchiba's original parallel commit, now even with master. 

Overview:
In nested sampling, the proposed points to replace a given live point are chosen independently from a set proposal distribution. This change supports parallelization by queuing up n points for evaluation on independent cores. ~~It also avoids wasting any parallel computations by evaluating all remaining proposed points in the queue before commencing with an update of the proposal distribution.~~ Some simple tests on a couple distributions show a reasonable speedup, but I haven't done any rigorous checks.

Changes/Comments:
- This implementation uses `ThreadPoolExecutor` to farm out jobs independently rather than rather than a simple `MPIPool` to submit them in batches. It's unclear whether the latter "simple" implementation would be any better/worse, but the former does lead to minimal changes to the overall structure of the code.
- Parallelization is currently only supported for the `'single'` and `'multi'` ellipsoid proposals. Parallelizing the `'classic'` MCMC-style run is straightforward but slightly more involved (so I haven't done it here).
- Although the `FakePool` and `FakeFuture` objects are a bit hacky, like @kbarbary I don't really see a much better option to deal with the serial case.
- Portions of the original proposal have been refactored into `propose_point` and `new_point`. The former grabs a valid proposal, while the latter checks whether the proposed likelihood satisfies the acceptance criteria.
- ~~The `new_point` and `Sampler.get_point_value` methods now also take a `should_update` flag. If this is set to `should_update=True`, the sampler does not automatically submit a new job to the queue. This allows the queue to empty out, ensuring we do not waste any function calls.~~

aarchiba commented on nestle.py:
```diff
@@ -856,16 +947,23 @@ def prior_transform(u):
+        pool = FakePool()
+    else:
+        if pool is None:
+            pool = concurrent.futures.ThreadPoolExecutor(n_threads)
```
ThreadPoolExedcutor is almost never of any use. Why not drop n_threads and simply require the user to supply a pool if they want parallelization? You do need an n_queue, unfortunately, which should be at least the number of cores that will be working.

joshspeagle replied:
I just thought it should be a backup in case users are lazy. I'd be happy to throw an error instead.

aarchiba replied:
I recommend throwing an error to avoid a stream of bug reports saying "parallelization doesn't make it any faster!"

kbarbary replied:
I agree with @aarchiba here. Ditch `n_threads` in favor of `queue_size` or `queue_length`, and throw an error if no pool is supplied.

Naming bikeshed: `n_foo` or `nfoo` denotes "number of foo" in the current parameter names (e.g., `npoints`, `ndim`). We only have one queue, so `n_queue` isn't quite right. Hence `queue_size` or `queue_length`.

joshspeagle replied:
Sounds good -- will do.

aarchiba commented:
I'm not so sure about the should_update flag. My original code was quite careful not to change the sequence of log_likelihood values used in any way; is it going to mess up the statistics to use log_likelihood values beyond the point where you would normally update the proposal distribution? Maybe it's better to throw those away; I don't remember the math of the algorithm well enough to know.

aarchiba commented on nestle.py:
```diff
@@ -560,13 +591,51 @@ class Sampler:
+        if not self.queue:
+            self.fill_queue()
+        x, v, f = self.queue.pop(0)
+        r = f.result()
```
If you're going to use points that have already been calculated even when you know you're going to update the proposal distribution, this is not the right way. You should go through the entire queue cancelling all the futures that should be cancelled (that is, any that haven't started yet) and either (a) taking values from only the ones that are ready now, or (b) wait for the ones that can't be cancelled (because they're running).

joshspeagle replied:
In terms of statistics, this change shouldn't affect the result in any way. The update schedule for the ellipsoids is purely an algorithmic choice. The number of iterations between ellipsoids is often (as in the default implementation) taken to be fixed, but it could easily be variable, set by some density criteria, etc.

joshspeagle replied:
I agree that sounds like the proper way to do things, but do you expect there to be a big change from this simplistic way of "emptying" the queue?

aarchiba replied:
The question is, what do you do with the points evaluated after the last valid one? You can throw them away or you can do more evaluations - not in parallel, presumably, unless you want to refill the queue and end up in the same situation. But it can take many tries to get a valid new point, and this could become the bottleneck. As long as it's statistically kosher, simply discarding them is the cleanest solution. should_update then still pulls any valid points out of what's left in the queue.

joshspeagle commented:
In terms of statistics, this change shouldn't affect the result in any way. The update schedule for the ellipsoids is purely an algorithmic choice. The number of iterations between ellipsoids is often (as in the default implementation) taken to be fixed, but it could easily be variable, set by some density criteria, etc.

aarchiba commented:
Don't forget to pull in the tests from my commit. Checking determinism would be nice too.

aarchiba commented on nestle.py:
```diff
@@ -560,13 +591,51 @@ class Sampler:
+
+        if not self.queue:
+            self.fill_queue()
+        x, v, f = self.queue.pop(0)
```
This will raise an exception if called with an empty queue and should_update=True. What should happen in this case? As far as I can tell the rest of the code expects to always be able to get a result from this function.

joshspeagle replied:
There are two checks to prevent this from happening. First, by default `get_point_value` attempts to fill the queue first, ensuring the queue is never empty. The second is after that updating the proposal distribution `should_update` is set to False.

aarchiba replied:
The problem is that new_point always needs to return an eligible point, so it might call into here with should_update even if the queue is empty.

aarchiba replied:
I think actually raising an exception here might be the cleanest way to get out of the new_point loop and handle the fact that there isn't actually a new point; catch it in the main loop and you're done. Just make sure it's an exception type that can't come from the user's function.

aarchiba commented:
The update schedule for the ellipsoids is an algorithmic choice that affects the result. So sneaking in a few free points is going to affect the result; your argument is just that it won't make it wrong. I think I'm convinced.

joshspeagle commented:
I'm happy to add in the modification to `runtests.py` file (or an equivalent) from your commit. Otherwise I don't think it's too hard for @kbarbary to make a quick change.

aarchiba commented on nestle.py:
```diff
@@ -655,19 +732,23 @@ def set_options(self, options):
-                    break
-            v = self.prior_transform(u)
-            logl = self.loglikelihood(v)
+            u, v, logl = self.get_point_value(should_update)
```
This calls get_point_value a potentially arbitrary number of times, but if should_update is set get_point_value will fail after the queue empties. If you want to make should_update work, you need to handle two possibilities: (a) there is an eligible point (logl>=loglstar) in the queue, or (b) there is no eligible point remaining. In fact you always need to handle (b), since sooner or later you'll run out of eligible points. In this case new_point needs to return some kind of special value and its caller needs to do something to take into account that no eligible point was found.

joshspeagle replied:
Good catch. So what's happening now then is the code gets to this point, calls `get_point_value` a bunch of times, and when it runs out fills up the queue again. This doesn't give the behavior I want. I think you can just solve this by putting an addition condition in the while loop -- I'll try that now.

joshspeagle commented:
As for the results, yes, you're correct -- the addition of a few extra points will affect the run compared to the vanilla implementation where this doesn't happen, but shouldn't inherently compromise nested sampling. 

As for the determinism, that's a good point -- it definitely would be worth a throwing out a few calls to keep that. I believe that since only the likelihood computations are farmed out, adding the additional points from the queue (especially in the simplistic implementation where I remove them by just not adding any more operations after passing the default iteration to update) shouldn't violate this, but I haven't tested this.

aarchiba commented on nestle.py:
```diff
@@ -920,13 +1018,18 @@ def prior_transform(u):
         # Choose a new point from within the likelihood constraint
         # (having logl > loglstar).
-        u, v, logl, nc = sampler.new_point(loglstar)
+        u, v, logl, nc = sampler.new_point(loglstar, should_update)
```
Here you have to handle the case where should_update is set and so no new eligible point was found. I think you can just use continue here, though maybe you should also increment it.

joshspeagle replied:
Yes, this is a concern. This problem doesn't end up occurring because of the bug you identified above but will be a concern if I fix things.

joshspeagle commented:
Okay -- I just removed the extra points modification and reverted back to the original implementation where extra points get tossed. Given that the computational gain was mostly minimal (a maximum extra n_threads proposals vs the large number of live points that get replaced between ellipsoid recalculations), I think the simpler modification makes more sense.

kbarbary commented:
Thanks for reviving this! I'm headed out of internet range at the moment, but I'll have a more detailed look on Monday.

kbarbary commented on nestle.py:
```diff
@@ -583,6 +651,13 @@ def update(self, pointvol):
 
+    def propose_point(self, u):
+        while True:
+            new_u = u + scale * self.ell.randoffset(rstate=self.rstate)
```
`scale` hasn't been passed to this function from `new_point`.

I think the original behavior is actually not correct: one shouldn't change the proposal distribution during the MCMC: a constant scale should be used during the entirety of `new_point`. I think it is fine to change the proposal distribution (using `scale`) between calls to `new_point` though. So perhaps `scale` should become an attribute of the class that is updated at the end of each `new_point` call.

joshspeagle replied:
Good catch. How would you like this to be implemented? It's straightforward to just add `scale` to `propose_point` and keep the functionality the same as the original class. Otherwise, I think the proper thing to do would be to adjust the acceptance ratio as already implemented, then once an "optimal" scale has been reached re-run the MCMC with `scale` fixed.

kbarbary replied:
But then wouldn't you have to add `scale` to `propose_point` in every Sampler subclass? (We don't need to fix the algorithm in this PR, but I thought it might help in getting around this problem.)

aarchiba replied:
I'm a little confused: is the original algorithm, without parallelization, wrong? That is, does it adjust `scale` too often? Or is this a bug introduced by the parallelization?

joshspeagle replied:
@kbarbary Not as far as I can tell, since this particular `propose_point` (and its use of `scale`) is already specific to the MCMC implementation.
@aarchiba I don't think the algorithm is "wrong", although it depends on what you're thinking of when you run the `classic` sampler. Ultimately, MCMC-like proposals are just a way to random-walk away from an original trial location to get an independent sample. As implemented, the proposal distribution changes over the course of a call to `new_point` to achieve a target acceptance ratio of 0.5. This violates detailed balance, but doesn't necessarily mean it's not an effective random walk. I haven't done any tests to see whether "fixing" this issue would improve things though.

aarchiba replied:
Actually as it stands you should get a NameError. I think since the ellipse is already stored in self, it won't hurt to put scale the too. But this emphasizes the need for tests.

kbarbary replied:
> Not as far as I can tell, since this particular propose_point (and its use of scale) is already specific to the MCMC implementation.

But the `Sampler` base class calls `propose_point()` in `fill_queue()`, so the interface has to be the same for all derived subclasses.

joshspeagle replied:
I have not implemented parallelism with the queue for the `'classic'` proposal, so right now it runs in serial as normal.

kbarbary replied:
Ah, I see. Why even separate out `propose_point()` for `ClassicSampler` then? can it just be inlined back into `ClassicSampler.new_point()` for the purpose of this PR?

(It's a bit weird to have all the queue stuff in the `Sampler.__init__` when `ClassicSampler` doesn't use it. It might be cleaner to have a derived class `EllipsoidSampler` that implements the queue stuff and then have `SingleEllipsoidSampler` and `MultiEllipsoidSampler` derive from that. But, not necessary for this PR. Just get the tests to pass. :wink: )

joshspeagle commented:
Okay -- I've now implemented the requested changes. I've confirmed that now the `'classic'` sampler now works regardless of the state of the queue/pool, so there hopefully shouldn't be any other problems.

kbarbary commented:
Looks like the travis failure is related to coverage not being set up correctly... the actual tests all pass. Thanks! I'll try to add the test from the original PR in the next few days.

exowanderer commented:
I see that this merge happened; but is there any documentation, notebooks, or examples to support it? I would very much like to use `nestle`; but I need it to be multi-processed.

In particular: I see that (with the github version; not the pip version) there is a kwarg for `pool`; but what are the inputs that `pool` is expecting? I tried:

1) `multiprocessing.Pool`
2) `multiprocessing.Pool(num_cores)`
3) `multiprocessing.Pool(num_cores).starmap`

None of them seemed to work or was otherwise unintuitive (the traceback was not related).

Thank you!

exowanderer commented:
If I proved `pool=multiprocessing.Pool(num_cores)`, then the traceback says

```python
In [9]: result = nestle.sample(loglike, prior_transform, 2, pool=Pool(8), queue_size=8)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-ce97228df1c0> in <module>()
----> 1 result = nestle.sample(loglike, prior_transform, ndims, pool=Pool(num_cores), queue_size=num_cores)

~/anaconda3/lib/python3.6/site-packages/nestle-0.2.0-py3.6.egg/nestle.py in sample(loglikelihood, prior_transform, ndim, npoints, method, update_interval, npdim, maxiter, maxcall, dlogz, decline_factor, rstate, callback, queue_size, pool, logl_args, logl_kwargs, prior_args, prior_kwargs, **options)
   1003 
   1004     # Initialize sampler
-> 1005     sampler.update(1./npoints)
   1006 
   1007     callback_info = {'it': 0,

~/anaconda3/lib/python3.6/site-packages/nestle-0.2.0-py3.6.egg/nestle.py in update(self, pointvol)
    704                                       minvol=True)
    705         self.ell.scale_to_vol(self.ell.vol * self.enlarge)
--> 706         self.fill_queue()
    707 
    708     def propose_point(self):

~/anaconda3/lib/python3.6/site-packages/nestle-0.2.0-py3.6.egg/nestle.py in fill_queue(self)
    622             x = self.propose_point()
    623             v = self.prior_transform(x)
--> 624             self.queue.append((x, v, self.pool.submit(self.loglikelihood, v)))
    625             self.submitted += 1
    626 

AttributeError: 'Pool' object has no attribute 'submit' <--**
```

I googled "python pool.submit" and found "concurrent.futures.ThreadPoolExecutor", which provide the following syntax -- and did not give me any errors.

```python
with futures.ThreadPoolExecutor(max_workers=2) as executor:
    result = nestle.sample(loglike, prior_transform, 2, pool=executor, queue_size=2)
```

Is this the proper use of the `pool` kwarg in `nestle.sample`?

kbarbary commented:
Yes, it does expect a `ThreadPoolExecutor` for the `pool` argument. Actually any `concurrent.futures.Executor` that fulfills the same interface will do. You're more likely to get a speed up with `ProcessPoolExecutor`: most likelihood functions are CPU- rather than IO-bound, so more threads doesn't help, but more processes does. But `multiprocessing.Pool` does *not* fulfill the interface.

It is very lightly documented in the docstring on master, but this all has not yet made it into a release version. https://github.com/kbarbary/nestle/blob/0741cdf9513f6477db2e645a101fc4a523ad36bc/nestle.py#L859-L861

It could use better docs and an example notebook!

I was hoping to have the time to clean up the implementation a bit and add docs before releasing a new verison, but I haven't had the time (due to a day job). If you are interested in contributing docs or examples or anything, let me know and I can help answer any questions you might have along the way.
This would probably prompt me to spend some more time on this as well. :)
